// ovsx/lib/util.js
"use strict";
/********************************************************************************
 * Copyright (c) 2019 TypeFox and others
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * SPDX-License-Identifier: EPL-2.0
 ********************************************************************************/
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs");
const tmp = require("tmp");
var util_1 = require("util");
exports.promisify = util_1.promisify;
function matchExtensionId(id) {
    return /^([\w\-]+)(?:\.|\/)([\w\-]+)$/.exec(id);
}
exports.matchExtensionId = matchExtensionId;
function optionalStat(path) {
    return new Promise((resolve, reject) => {
        fs.stat(path, (err, stats) => resolve(stats));
    });
}
exports.optionalStat = optionalStat;
function makeDirs(path) {
    return new Promise((resolve, reject) => {
        if (fs.existsSync(path)) {
            resolve();
        }
        else {
            fs.mkdir(path, { recursive: true }, err => {
                if (err)
                    reject(err);
                else
                    resolve();
            });
        }
    });
}
exports.makeDirs = makeDirs;
function createTempFile(options) {
    return new Promise((resolve, reject) => {
        tmp.tmpName(options, (err, name) => {
            if (err)
                reject(err);
            else
                resolve(name);
        });
    });
}
exports.createTempFile = createTempFile;
function handleError(debug) {
    return reason => {
        if (reason instanceof Error && !debug) {
            console.error(`\u274c  ${reason.message}`);
        }
        else if (typeof reason === 'string') {
            console.error(`\u274c  ${reason}`);
        }
        else if (reason !== undefined) {
            console.error(reason);
        }
        else {
            console.error('An unknown error occurred.');
        }
        process.exit(1);
    };
}
exports.handleError = handleError;
function statusError(response) {
    if (response.statusMessage)
        return new Error(`The server responded with status ${response.statusCode}: ${response.statusMessage}`);
    else
        return new Error(`The server responded with status ${response.statusCode}.`);
}
exports.statusError = statusError;
