// @graphql-tools/url-loader/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopNamespace(e) {
    if (e && e.__esModule) { return e; } else {
        var n = {};
        if (e) {
            Object.keys(e).forEach(function (k) {
                var d = Object.getOwnPropertyDescriptor(e, k);
                Object.defineProperty(n, k, d.get ? d : {
                    enumerable: true,
                    get: function () {
                        return e[k];
                    }
                });
            });
        }
        n['default'] = e;
        return n;
    }
}

const tslib = require('tslib');
const graphql = require('graphql');
const utils = require('@graphql-tools/utils');
const validUrl = require('valid-url');
const crossFetch = require('cross-fetch');
const wrap = require('@graphql-tools/wrap');
const subscriptionsTransportWs = require('subscriptions-transport-ws');
const websocket = require('websocket');

/**
 * This loader loads a schema from a URL. The loaded schema is a fully-executable,
 * remote schema since it's created using [@graphql-tools/wrap](/docs/remote-schemas).
 *
 * ```
 * const schema = await loadSchema('http://localhost:3000/graphql', {
 *   loaders: [
 *     new UrlLoader(),
 *   ]
 * });
 * ```
 */
class UrlLoader {
    loaderId() {
        return 'url';
    }
    canLoad(pointer, options) {
        return tslib.__awaiter(this, void 0, void 0, function* () {
            return this.canLoadSync(pointer, options);
        });
    }
    canLoadSync(pointer, _options) {
        return !!validUrl.isWebUri(pointer);
    }
    buildAsyncExecutor({ pointer, fetch, extraHeaders, defaultMethod, useGETForQueries, }) {
        const HTTP_URL = pointer.replace('ws:', 'http:').replace('wss:', 'https:');
        return ({ document, variables }) => tslib.__awaiter(this, void 0, void 0, function* () {
            let method = defaultMethod;
            if (useGETForQueries) {
                method = 'GET';
                for (const definition of document.definitions) {
                    if (definition.kind === graphql.Kind.OPERATION_DEFINITION) {
                        if (definition.operation !== 'query') {
                            method = defaultMethod;
                        }
                    }
                }
            }
            let fetchResult;
            const query = graphql.print(document);
            switch (method) {
                case 'GET':
                    const urlObj = new URL(HTTP_URL);
                    urlObj.searchParams.set('query', query);
                    if (variables && Object.keys(variables).length > 0) {
                        urlObj.searchParams.set('variables', JSON.stringify(variables));
                    }
                    const finalUrl = urlObj.toString();
                    fetchResult = yield fetch(finalUrl, {
                        method: 'GET',
                        headers: extraHeaders,
                    });
                    break;
                case 'POST':
                    fetchResult = yield fetch(HTTP_URL, {
                        method: 'POST',
                        body: JSON.stringify({
                            query,
                            variables,
                        }),
                        headers: extraHeaders,
                    });
                    break;
            }
            return fetchResult.json();
        });
    }
    buildSubscriber(pointer, webSocketImpl) {
        const WS_URL = pointer.replace('http:', 'ws:').replace('https:', 'wss:');
        const subscriptionClient = new subscriptionsTransportWs.SubscriptionClient(WS_URL, {}, webSocketImpl);
        return ({ document, variables }) => tslib.__awaiter(this, void 0, void 0, function* () {
            return utils.observableToAsyncIterable(subscriptionClient.request({
                query: document,
                variables,
            }));
        });
    }
    getExecutorAndSubscriber(pointer, options) {
        return tslib.__awaiter(this, void 0, void 0, function* () {
            let headers = {};
            let fetch = crossFetch.fetch;
            let defaultMethod = 'POST';
            let webSocketImpl = websocket.w3cwebsocket;
            if (options) {
                if (Array.isArray(options.headers)) {
                    headers = options.headers.reduce((prev, v) => (Object.assign(Object.assign({}, prev), v)), {});
                }
                else if (typeof options.headers === 'object') {
                    headers = options.headers;
                }
                if (options.customFetch) {
                    if (typeof options.customFetch === 'string') {
                        const [moduleName, fetchFnName] = options.customFetch.split('#');
                        fetch = yield new Promise(function (resolve) { resolve(_interopNamespace(require(moduleName))); }).then(module => (fetchFnName ? module[fetchFnName] : module));
                    }
                    else {
                        fetch = options.customFetch;
                    }
                }
                if (options.webSocketImpl) {
                    if (typeof options.webSocketImpl === 'string') {
                        const [moduleName, webSocketImplName] = options.webSocketImpl.split('#');
                        webSocketImpl = yield new Promise(function (resolve) { resolve(_interopNamespace(require(moduleName))); }).then(module => webSocketImplName ? module[webSocketImplName] : module);
                    }
                    else {
                        webSocketImpl = options.webSocketImpl;
                    }
                }
                if (options.method) {
                    defaultMethod = options.method;
                }
            }
            const extraHeaders = Object.assign({ Accept: 'application/json', 'Content-Type': 'application/json' }, headers);
            const executor = this.buildAsyncExecutor({
                pointer,
                fetch,
                extraHeaders,
                defaultMethod,
                useGETForQueries: options.useGETForQueries,
            });
            let subscriber;
            if (options.enableSubscriptions) {
                subscriber = this.buildSubscriber(pointer, webSocketImpl);
            }
            return {
                executor,
                subscriber,
            };
        });
    }
    getSubschemaConfig(pointer, options) {
        return tslib.__awaiter(this, void 0, void 0, function* () {
            const { executor, subscriber } = yield this.getExecutorAndSubscriber(pointer, options);
            return {
                schema: yield wrap.introspectSchema(executor, undefined, options),
                executor,
                subscriber,
            };
        });
    }
    load(pointer, options) {
        return tslib.__awaiter(this, void 0, void 0, function* () {
            const subschemaConfig = yield this.getSubschemaConfig(pointer, options);
            const remoteExecutableSchema = wrap.wrapSchema(subschemaConfig);
            return {
                location: pointer,
                schema: remoteExecutableSchema,
            };
        });
    }
    loadSync() {
        throw new Error('Loader Url has no sync mode');
    }
}

exports.UrlLoader = UrlLoader;
